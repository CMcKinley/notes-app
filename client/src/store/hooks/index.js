export { default as useGetNotes } from './useGetNotes';
export { default as useEditNotes } from './useEditNotes';
export { default as useAddNotes } from './useAddNotes';