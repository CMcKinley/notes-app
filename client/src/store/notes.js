import { handleActions, createAction } from "redux-actions";
import cloneDeep from "lodash.clonedeep";

const defultState = {
  noteIds: [],
  noteMap: {},
  noteToEdit: null
};

const prefix = "NOTES/";

// Action Creators
export const setNotes = createAction(`${prefix}SET_NOTES`);
export const editNoteInStore = createAction(`${prefix}EDIT_NOTE`);
export const addNoteToStore = createAction(`${prefix}ADD_NOTE`);
export const setNoteToEdit = createAction(`${prefix}SET_NOTE_TO_EDIT`);

// Reducer
export default handleActions(
  {
    // Load notes into state. Notes are stored as an array of ids and an object where each key is the note's id and the value is the note object itself.
    [setNotes]: (state, { payload }) => {
      const notes = payload || [];
      const noteIds = [];
      const noteMap = {};

      notes.forEach(note => {
        const { id } = note || {};
        noteIds.push(id);
        noteMap[id] = note;
      });

      return { ...state, noteIds, noteMap };
    },

    // Edit an existing note
    [editNoteInStore]: (state, { payload }) => {
      const { id } = payload || {};
      // The values of the note map are objects so we need to clone deep the noteMap to ensure each object's properties are not references to the original
      const newMap = cloneDeep(state.noteMap);
      newMap[id] = { ...newMap[id], ...payload };
      return { ...state, noteMap: newMap };
    },

    // Add a new note to the id array and note map
    [addNoteToStore]: (state, { payload }) => {
      const { id } = payload || {};
      const noteIds = [...state.noteIds];
      // The values of the note map are objects so we need to clone deep the noteMap to ensure each object's properties are not references to the original
      const noteMap = cloneDeep(state.noteMap);
      // Add note to beginning of array
      noteIds.unshift(payload.id);
      noteMap[id] = payload;
      return { ...state, noteIds, noteMap };
    },

    // Set the note that is to be edited in store
    [setNoteToEdit]: (state, { payload }) => ({ ...state, noteToEdit: payload })
  },
  defultState
);
