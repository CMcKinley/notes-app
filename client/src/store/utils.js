import axios from "axios";

// Store tools will be used to make http calls to the node api
export default class StoreTools {
  async base(url, body, method = "get", headers = {}) {
    let config = { url: `/api/v1/${url}`, method, headers };
    if (body) {
      config.data = body;
    }
    return axios(config);
  }
  // GET requests
  async get(url, headers = {}) {
    return this.base(url, null, "get", headers);
  }
  // PUT requests
  put(url, body, headers = {}) {
    return this.base(url, body, "put", headers);
  }
  // POST requests
  post(url, body, headers = {}) {
    return this.base(url, body, "post", headers);
  }
}
