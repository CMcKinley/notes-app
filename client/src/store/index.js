import { combineReducers, createStore, applyMiddleware } from "redux";
import thunk from "redux-thunk";
import { composeWithDevTools } from "redux-devtools-extension";
import notes from "./notes.js";

const store = combineReducers({ notes });

const composeEnhancers = composeWithDevTools({});

export default createStore(store, composeEnhancers(applyMiddleware(thunk)));
