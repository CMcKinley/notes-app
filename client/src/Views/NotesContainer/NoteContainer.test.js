import React from "react";
import renderer from "react-test-renderer";
import NoteContainer from "./index.js";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import { defultState } from "store/notes";
export const mockStore = configureStore([]);

const initialState = { notes: defultState  };
const store = mockStore(initialState);

it("Renders correctly", () => {
  const component = renderer.create(
    <Provider store={store}>
      <NoteContainer />
    </Provider>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
