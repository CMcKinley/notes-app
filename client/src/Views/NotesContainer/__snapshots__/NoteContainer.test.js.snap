// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly 1`] = `
<section>
  <div
    className="makeStyles-listContainer-1"
  >
    <ul
      className="MuiList-root MuiList-padding"
    />
  </div>
</section>
`;
